README.md
Mappy
====

## Документация
[ru Google.Maps](https://github.com/Niklan/Mappy/wiki/ru_Google.Maps)
[ru Yandex.Maps](https://github.com/Niklan/Mappy/wiki/ru_Yandex.Maps)

CHANGELOG.txt
7.x-1.2
-----------------
+ Clusters for Google.Maps.
+ Support multiply maps on the same page.
+ New method of insertion maps. Token (filter) [mappy:service address:''].
+ Administration interface.
+ Added default value for zoom, width and height (if user don't set).
+ All JS's reworked.
+ Google JSAPI load on when it needed.
+ Now you can chose which page load Mappy.js.
+ Attributes no more case sensitive.

7.x-1.2
-----------------
+ Added a new option for the scrollwheel google maps.
+ By xandeadx: Added new option type to google maps.
+ Fixed a bug where a tpl template files needed to write google, instead mappy: google. Now run both.

js/mappy.js
/**
 * @file
 * This is main script of Mappy. His route other js files.
 */

(function ($) {
    $(document).ready(function () {
        // If found mappy:yandex tag, we attach Yandex Maps script.
        if ($("yandex").length || $("mappy\\:yandex").length) {
            // First we connect Yandex.Maps API script.
            $.getScript("http://api-maps.yandex.ru/2.0/?load=package.full&lang=ru-RU", function () {
                // If script loaded, then load our script.
                $.getScript("/" + Drupal.settings.mappy_location + "/js/yandex.mappy.js");
            });
        }

        // If found mappy:google tag, we attach Google Maps script.
        if ($("google").length || $("mappy\\:google").length) {
            // First, we load main Google script.
            $.getScript("https://www.google.com/jsapi", function () {
                // If enabled clusters, we must load additional library.
                if ($("google, mappy\\:google").attr("clusters")) {
                    $.getScript("http://google-maps-utility-library-v3.googlecode.com/svn/trunk/markerclusterer/src/markerclusterer.js");
                }
                // Finally, we load our script for handle tag.
                $.getScript("/" + Drupal.settings.mappy_location + "/js/google.mappy.js");
            });
        }
    });
})(jQuery);


js/yandex.mappy.js
/**
 * @file
 * This is script for Yandex.Maps 2.0 (old version).
 *
 * We use get(0).getAttribute instead of attr() because Drupal 7 jquery is
 * version 1.4. In this version attr() case sensitive, we don't need this.
 */

(function ($) {
    // Just optimisation for clear reading code.
    var mappy_tag = $("yandex, mappy\\:yandex");
    // Create own array for storage information for each map (from tag).
    var mappy = [];
    // Create array of maps.
    var maps = []
    // We apply map for each tag founded on page.
    mappy_tag.each(function (index) {
        // Instance of tag.
        var mappy_instance = $(this);
        mappy[index] = {
            // Address of map.
            address: mappy_instance.get(0).getAttribute("address").split(";"),
            // Lat. & long. center of map. We set a little bit later.
            centerCoordinates: "",
            // Width of the map. If not present, we get default value from settings.
            width: mappy_instance.get(0).getAttribute("width") > 0 ? mappy_instance.get(0).getAttribute("width") : Drupal.settings.mappy_yandex_width,
            // Height of the map. If not present, we get default value from settings.
            height: mappy_instance.get(0).getAttribute("height") > 0 ? mappy_instance.get(0).getAttribute("height") : Drupal.settings.mappy_yandex_height,
            // Scale of the map.
            zoom: mappy_instance.get(0).getAttribute("zoom") > 0 ? parseInt(mappy_instance.get(0).getAttribute("zoom")) : 17,
            // Zoom button (X,Y).
            zoomControl: mappy_instance.get(0).getAttribute("zoomControl") ? mappy_instance.get(0).getAttribute("zoomControl").split(",") : false,
            // Small zoom button (X,Y).
            smallZoomControl: mappy_instance.get(0).getAttribute("smallZoomControl") ? mappy_instance.get(0).getAttribute("smallZoomControl").split(",") : false,
            // Disable balloons.
            addressPlacemark: (mappy_instance.get(0).getAttribute("addressPlacemark") == "false") ? false : true,
            // If TRUE: users can choose layers.
            mapTypeControl: mappy_instance.get(0).getAttribute("mapTypeControl") == "true" ? true : false,
            // If TRUE: display button for controlling jams on road.
            mapTrafficControl: mappy_instance.get(0).getAttribute("mapTrafficControl") == "true" ? true : false,
            // Content for balloons.
            balloonContent: (mappy_instance.attr("balloonContent")) ? mappy_instance.attr("balloonContent").split(";") : false,
            // Route controllers.
            route: mappy_instance.get(0).getAttribute("route") ? mappy_instance.get(0).getAttribute("route").split(",") : false,
            // If TRUE: balloons will merge into one big, before they zoomed enough for seeing separately.
            cluster: mappy_instance.get(0).getAttribute("clusters") == "true" ? true : false
        };
        console.log(mappy);
        // Obtain the coordinates of the first address (for map center).
        $.ajax({
            url: 'http://geocode-maps.yandex.ru/1.x/?format=json&geocode=' + mappy[index].address[0] + '&result=1',
            success: function (data) {
                mappy[index].centerCoordinates = data.response.GeoObjectCollection.featureMember[0].GeoObject.Point.pos.split(" ").reverse();
                create_map();
            }
        });

        // Generation of map.
        function create_map() {
            // Add id tag for apply map.
            mappy_instance.attr('id', 'mappy-yandex-' + [index]);
            mappy_instance.attr('class', 'mappy yandex');
            // Add some ccs for map.
            mappy_instance.css({
                'display': 'block',
                'width': mappy[index].width,
                'height': mappy[index].height
            });

            // Class for multiple geocoding.
            function MultiGeocoder(options) {
                this._options = options || {};
            }

            // Multigeocoding.
            MultiGeocoder.prototype.geocode = function (requests, options) {
                var self = this,
                    opts = ymaps.util.extend({}, self._options, options),
                    size = requests.length,
                    promise = new ymaps.util.Promise(),
                    result = [],
                    geoObjects = new ymaps.GeoObjectArray();

                requests.forEach(function (request, index) {
                    ymaps.geocode(request, opts).then(
                        function (response) {
                            var geoObject = response.geoObjects.get(0);

                            geoObject && (result[index] = geoObject);
                            --size || (result.forEach(geoObjects.add, geoObjects), promise.resolve({ geoObjects: geoObjects }));
                        },
                        function (err) {
                            promise.reject(err);
                        }
                    );
                });

                return promise;
            };

            // Prepare variable for our map.
            maps[index];

            // Wait until download API. _mm._sc
            ymaps.ready(init);

            function init() {
                // Create a new instance of the map.
                maps[index] = [];
                maps[index].map = new ymaps.Map('mappy-yandex-' + [index], {
                    // Coordinates of the center of the map.
                    center: mappy[index].centerCoordinates,
                    // The scale of the map.
                    zoom: mappy[index].zoom
                });

                // Add balloon on map;
                function addBalloon(balloonCoordinates, balloonAddress, balloonContent) {
                    var coords = balloonCoordinates;
                    var address = balloonAddress;
                    var content = balloonContent;
                    balloon = new ymaps.Placemark(coords, {
                        balloonContent: content,
                        hintContent: address
                    }, {
                        preset: 'twirl#blueIcon'
                    });
                    maps[index].map.geoObjects
                        .add(balloon);
                }

                // Array Geocoding addresses and coordinates, as well as adding labels.
                if (mappy[index].address.length > 0) {
                    var geocoding_coord;
                    for (var i = 0; i < mappy[index].address.length; i++) {
                        // Prepare balloon content.
                        if (mappy[index].balloonContent[i] != "undefined") {
                            var content = mappy[index].balloonContent[i];
                        }

                        $.ajax({
                            url: 'http://geocode-maps.yandex.ru/1.x/?format=json&geocode=' + mappy[index].address[i] + '&result=1',
                            success: function (data) {
                                geocoding_coord = data.response.GeoObjectCollection.featureMember[0].GeoObject.Point.pos.split(" ").reverse();
                                var address = data.response.GeoObjectCollection.featureMember[0].GeoObject.name;
                                (mappy[index].addressPlacemark) ? addBalloon(geocoding_coord, address, content) : false;
                            },
                            async: false
                        });
                    }
                }

                // Zoom control (X,Y).
                if (mappy[index].zoomControl) {
                    maps[index].map.controls
                        .add('zoomControl', { left: mappy[index].zoomControl[0], top: mappy[index].zoomControl[1] });
                }

                // Small zoom control (X,Y).
                if (mappy[index].smallZoomControl) {
                    maps[index].map.controls
                        .add('smallZoomControl', { left: mappy[index].smallZoomControl[0], top: mappy[index].smallZoomControl[1] });
                }

                // Users can select layers.
                if (mappy[index].mapTypeControl) {
                    maps[index].map.controls
                        .add('typeSelector');
                }

                // Button showing the traffic jams on the road.
                if (mappy[index].mapTrafficControl) {
                    maps[index].map.controls
                        .add(new ymaps.control.TrafficControl());
                }

                // Create a route.
                if (mappy[index].route) {
                    var router;
                    $("#" + mappy[index].route[0]).click(function () {
                        // Address 'from'.
                        var route_address = $("#" + mappy[index].route[1]).val();
                        // End of route.
                        var address_coordinates = mappy[index].centerCoordinates;

                        // Generate route.
                        ymaps.route([
                            // Form.
                            route_address,
                            // Where.
                            address_coordinates
                        ], {
                            // Autozooming.
                            mapStateAutoApply: true
                        }).then(function (route) {
                                // Clear previous route.
                                if (router) {
                                    maps[index].map.geoObjects
                                        .remove(router);
                                }
                                // Route from API.
                                router = route;
                                // Draw route.
                                maps[index].map.geoObjects.add(router);
                                // Get route points .
                                var points = route.getWayPoints();
                                // Bubble type.
                                points.options.set('preset', 'twirl#redStretchyIcon');
                                // Icon and text of 'From' bubble.
                                points.get(0).properties.set('iconContent', 'Ваше местоположение');
                                // Icon and text of 'Where' bubble.
                                points.get(1).properties.set('iconContent', 'Мы здесь!');
                            }, function (error) {
                                // Display error if exept.
                                alert("Error: " + error.message);
                            });
                    });
                }
            }
        }
    });


})(jQuery);


js/google.mappy.js
/**
 * @file
 * This is script for Google.Maps.
 *
 * We use get(0).getAttribute instead of attr() because Drupal 7 jquery is
 * version 1.4. In this version attr() case sensitive, we don't need this.
 */

(function ($) {
    // Call Google.Maps API v3.
    google.load("maps", "3", {other_params: 'sensor=false', callback: function () {
        // Just optimisation for clear reading code.
        var mappy_tag = $("google, mappy\\:google");
        // Create own array for storage information for each map (from tag).
        var mappy = [];
        // Create array of maps.
        var maps = []
        // We apply map for each tag founded on page.
        mappy_tag.each(function (index) {
            // Instance of tag.
            var mappy_instance = $(this);
            mappy[index] = {
                // Address of map.
                address: mappy_instance.get(0).getAttribute("address").split(";"),
                // Lat. & long. center of map. We set a little bit later.
                center_lat: "",
                center_lng: "",
                // Width of the map. If not present, we get default value from settings.
                width: mappy_instance.get(0).getAttribute("width") > 0 ? mappy_instance.get(0).getAttribute("width") : Drupal.settings.mappy_google_width,
                // Height of the map. If not present, we get default value from settings.
                height: mappy_instance.get(0).getAttribute("height") > 0 ? mappy_instance.get(0).getAttribute("height") : Drupal.settings.mappy_google_height,
                // Scale of the map.
                zoom: mappy_instance.get(0).getAttribute("zoom") > 0 ? parseInt(mappy_instance.get(0).getAttribute("zoom")) : 17,
                // The map type.
                type: mappy_instance.get(0).getAttribute("type") ? mappy_instance.get(0).getAttribute("type") : "ROADMAP",
                // If TRUE: display zoom control.
                zoomControl: mappy_instance.get(0).getAttribute("zoomControl") == "false" ? false : true,
                // If TRUE: allow users to use street view.
                streetViewControl: mappy_instance.get(0).getAttribute("streetViewControl") == "false" ? false : true,
                // If TRUE: allow users to select map layer.
                mapTypeControl: mappy_instance.get(0).getAttribute("mapTypeControl") == "false" ? false : true,
                // If TRUE: display circle for move the map.
                panControl: mappy_instance.get(0).getAttribute("panControl") == "false" ? false : true,
                // Content for balloons.
                balloonContent: (mappy_instance.attr("balloonContent")) ? mappy_instance.attr("balloonContent").split(";") : false,
                // If FALSE: 'scrolling' page, instead of map zooming.
                scrollWheel: mappy_instance.get(0).getAttribute("scrollwheel") == "false" ? false : true,
                // If TRUE: balloons will merge into one big, before they zoomed enough for seeing separately.
                cluster: mappy_instance.get(0).getAttribute("clusters") == "true" ? true : false
            };

            // Obtain the coordinates of the first address (for map center).
            $.ajax({
                url: 'http://maps.googleapis.com/maps/api/geocode/json?address=' + mappy[index].address + '&sensor=false',
                success: function (data) {
                    mappy[index].center_lat = data.results[0].geometry.location.lat;
                    mappy[index].center_lng = data.results[0].geometry.location.lng;
                    initialize();
                },
                async: false
            });

            // Prepare variable for our map.
            maps[index];

            // Initialization of the map.
            function initialize() {
                // Add id tag for apply map.
                mappy_instance.attr('id', 'mappy-google-' + [index]);
                mappy_instance.attr('class', 'mappy google');
                // Add some ccs for map.
                mappy_instance.css({
                    'display': 'block',
                    'width': mappy[index].width,
                    'height': mappy[index].height
                });

                // Map options.
                maps[index] = [];
                maps[index].mapOptions = {
                    zoom: mappy[index].zoom,
                    zoomControl: mappy[index].zoomControl,
                    mapTypeControl: mappy[index].mapTypeControl,
                    streetViewControl: mappy[index].streetViewControl,
                    panControl: mappy[index].panControl,
                    center: new google.maps.LatLng(mappy[index].center_lat, mappy[index].center_lng),
                    mapTypeId: google.maps.MapTypeId[mappy[index].type],
                    scrollwheel: mappy[index].scrollWheel
                };

                // Create a map.
                maps[index].map = new google.maps.Map(document.getElementById('mappy-google-' + [index]), maps[index].mapOptions);

                // Add markers on the map.
                if (mappy_instance.attr("addressPlacemark") != "false") {
                    // We use markers only for cluster for now.
                    var markers = [];
                    for (var i = 0; i < mappy[index].address.length; i++) {
                        var placemarkLat;
                        var placemarkLng;
                        var placemarkAddress;
                        var placemarkContent = mappy[index].balloonContent[i];

                        $.ajax({
                            url: 'http://maps.googleapis.com/maps/api/geocode/json?address=' + mappy[index].address[i] + '&sensor=false',
                            success: function (data) {
                                placemarkLat = data.results[0].geometry.location.lat;
                                placemarkLng = data.results[0].geometry.location.lng;
                                placemarkAddress = data.results[0].address_components[1].long_name + " " + data.results[0].address_components[0].long_name;
                            },
                            async: false
                        });

                        // Create marker.
                        var marker = new google.maps.Marker({
                            position: new google.maps.LatLng(placemarkLat, placemarkLng),
                            map: maps[index].map,
                            title: placemarkAddress
                        });
                        markers.push(marker);

                        // Add balloon with content.
                        function addInfoWindow(marker, message) {
                            var info = message;

                            var infoWindow = new google.maps.InfoWindow({
                                content: message
                            });

                            google.maps.event.addListener(marker, 'click', function () {
                                infoWindow.open(maps[index].map, marker);
                            });
                        }

                        addInfoWindow(marker, placemarkContent);
                    }

                    if (mappy[index].cluster) {
                        var markerCluster = new MarkerClusterer(maps[index].map, markers);
                    }
                }
            }
        });
    }});
})(jQuery);
